lib/krikri/harvesters/ia_harvester.rb

```diff
@@ -0,0 +1,160 @@
+require 'uri'
+require 'json'
+require 'async_uri_getter'
+
+module Krikri::Harvesters
+  ##
+  # A harvester implementation for Internet Archive
+  #
+  class IaHarvester
+    include Krikri::Harvester
+
+    DEFAULT_THREAD_COUNT = 10
+    DEFAULT_MAX_RECORDS = 0
+
+    DOWNLOAD_BASE_URI = 'http://archive.org/download'
+
+    ##
+    # Initialize, and set default options as appropriate.
+    #
+    # @param opts [Hash] a hash of options as defined by {.expected_opts}
+    #
+    # @example
+    #    Typical instantiation, good for most cases:
+    #
+    #      Krikri::Harvesters::IaHarvester.new(uri: 'http://archive.org/...')
+    #
+    # Accepts options passed as `:ia => opts`
+    #
+    # Options allowed are:
+    #
+    #   - threads:     The number of records to fetch asynchronously
+    #                  in a batch (default: 10)
+    #   - name:        See Krikri::Harvester#initialize.  Defaults to "ia"
+    #   - max_records: The maximum number of records to harvest
+    #                  0 means no limit (default 0)
+    #
+    def initialize(opts = {})
+      @opts = opts.fetch(:ia, {})
+      super
+
+      @opts[:threads] ||= DEFAULT_THREAD_COUNT
+      @opts[:max_records] ||= DEFAULT_MAX_RECORDS
+
+      @http = AsyncUriGetter.new(opts: { follow_redirects: true })
+    end
+
+    ##
+    # @see Krikri::Harvester.expected_opts
+    def self.expected_opts
+      {
+        key: :ia,
+        opts: {
+          threads:  { type: :integer, required: false },
+          max_records:  { type: :integer, required: false }
+        }
+      }
+    end
+
+    ##
+    # @see Krikri::Harvester#count
+    def count
+      Integer(collection_search['response']['numFound'])
+    end
+
+    ##
+    # @return [Enumerator::Lazy] an enumerator of the records targeted by this
+    #   harvester.
+    def records
+      threads = @opts.fetch(:threads)
+      max_records = @opts.fetch(:max_records)
+      last_record = max_records == 0 ? count : [count, max_records].min
+
+      (0...last_record - 1).step(threads).lazy.flat_map do |offset|
+        enumerate_records(record_ids(start: offset, rows: threads))
+      end
+    end
+
+    ##
+    # @param identifier [#to_s] the identifier of the record to get
+    # @return [#to_s] the record
+    def get_record(identifier)
+      enumerate_records([identifier]).first
+    end
+
+    private
+
+    ##
+    # Get a batch of records
+    # @param identifiers [Array] identifiers for the docs to get
+    # @return [Array] an array of @record_class instances
+    def enumerate_records(identifiers)
+      batch = []
+      identifiers.each do |id|
+        meta_uri = "#{DOWNLOAD_BASE_URI}/#{id}/#{id}_meta.xml"
+
+        batch << { :meta_request => @http.add_request(uri: URI.parse(meta_uri)),
+                   :id => id }
+      end
+
+      batch.each { |r| r[:meta_request].join }
+
+      batch.each do |record|
```

Comment:
Can we add in a request to fetch the `*_files.xml` for the item as well? (see [here](https://github.com/dpla/ingestion/blob/develop/lib/fetchers/ia_fetcher.py#L271) for instance)